File: app/src/main/java/cz/marvincz/rssnotifier/xml/XmlRssChannelConverter.java
package cz.marvincz.rssnotifier.xml;

import android.net.Uri;

import org.apache.commons.lang3.reflect.TypeLiteral;
import org.apache.commons.lang3.reflect.Typed;

import java.util.Arrays;
import java.util.List;

import androidx.annotation.Nullable;
import cz.marvincz.rssnotifier.model.ChannelWithItems;
import cz.marvincz.rssnotifier.model.RssItem;
import cz.marvincz.xmlpullparserconverter.XmlComplexConverter;
import cz.marvincz.xmlpullparserconverter.XmlFieldDefinition;

public class XmlRssChannelConverter extends XmlComplexConverter<ChannelWithItems> {
    public XmlRssChannelConverter() {
        super(t -> new ChannelWithItems());
    }

    @Override
    public Typed<ChannelWithItems> getType() {
        return new TypeLiteral<ChannelWithItems>() {};
    }

    @Nullable
    @Override
    protected List<XmlFieldDefinition<ChannelWithItems, ?>> getAttributes() {
        return null;
    }

    @Nullable
    @Override
    protected List<XmlFieldDefinition<ChannelWithItems, ?>> getTags() {
        return Arrays.asList(
                new XmlFieldDefinition<>("title", (ch, t) -> ch.title = t, new TypeLiteral<String>() {}),
                new XmlFieldDefinition<>("description", (ch, d) -> ch.description = d, new TypeLiteral<String>() {}),
                new XmlFieldDefinition<>("link", (ch, l) -> ch.link = l, new TypeLiteral<Uri>() {}),
                new XmlFieldDefinition<>("item", (ch, i) -> ch.items = i, new TypeLiteral<List<RssItem>>() {}));
    }
}

File: app/src/main/java/cz/marvincz/rssnotifier/xml/XmlRssChannelConverter.kt
package cz.marvincz.rssnotifier.xml

import cz.marvincz.rssnotifier.model.RssChannel
import cz.marvincz.xmlpullparserconverter.FieldValueSetter
import cz.marvincz.xmlpullparserconverter.XmlComplexConverter
import cz.marvincz.xmlpullparserconverter.XmlFieldDefinition
import org.apache.commons.lang3.reflect.TypeLiteral

class XmlRssChannelConverter : XmlComplexConverter<RssChannel>({ RssChannel() }) {

    override fun getType() = object : TypeLiteral<RssChannel>() {}

    override fun getAttributes() = null

    override fun getTags() = arrayListOf(
                XmlFieldDefinition<RssChannel, String>("title", FieldValueSetter { ch, t -> ch.title = t }, object : TypeLiteral<String>() {}),
                XmlFieldDefinition<RssChannel, String>("description", FieldValueSetter { ch, d -> ch.description = d }, object : TypeLiteral<String>() {}),
                XmlFieldDefinition<RssChannel, String>("link", FieldValueSetter { ch, l -> ch.link = l }, object : TypeLiteral<String>() {})
    )
}
